addOns/automation/src/main/java/org/zaproxy/addon/automation/tests/UrlPresenceTest.java

```diff
@@ -0,0 +1,406 @@
+                boolean requestheaderfound = false;
+                boolean requestbodyfound = false;
+                boolean responseheaderfound = false;
+                boolean responsebodyfound = false;
+                boolean result = false;
+
+                if (isNullOrEmpty(requestHeaderRegex)
+                        && isNullOrEmpty(requestBodyRegex)
+                        && isNullOrEmpty(responseHeaderRegex)
+                        && isNullOrEmpty(responseBodyRegex)) {
+
+                    if (onFail.equals(OnFail.ERROR)) {
+                        progress.error(
+                                Constant.messages.getString(
+                                        "automation.tests.url.noregex",
+                                        getJobType(),
+                                        getName(),
+                                        getTestType()));
+                    } else {
+                        progress.warn(
+                                Constant.messages.getString(
+                                        "automation.tests.url.noregex",
+                                        getJobType(),
+                                        getName(),
+                                        getTestType()));
+                    }
+                    return false;
+                }
+
+                if (!isNullOrEmpty(requestHeaderRegex)) {
+                    requestheaderfound =
+                            matchPattern(
+                                    requestHeaderRegex, msg, PARAM_REQUEST_HEADER_REGEX, progress);
+                } else if (operator.equalsIgnoreCase("and")) {
+                    requestheaderfound = true;
+                }
+
+                if (!isNullOrEmpty(requestBodyRegex)) {
+                    requestbodyfound =
+                            matchPattern(requestBodyRegex, msg, PARAM_REQUEST_BODY_REGEX, progress);
+                } else if (operator.equalsIgnoreCase("and")) {
+                    requestbodyfound = true;
+                }
+
+                if (!isNullOrEmpty(responseHeaderRegex)) {
+                    responseheaderfound =
+                            matchPattern(
+                                    responseHeaderRegex,
+                                    msg,
+                                    PARAM_RESPONSE_HEADER_REGEX,
+                                    progress);
+                } else if (operator.equalsIgnoreCase("and")) {
+                    responseheaderfound = true;
+                }
+
+                if (!isNullOrEmpty(responseBodyRegex)) {
+                    responsebodyfound =
+                            matchPattern(
+                                    responseBodyRegex, msg, PARAM_RESPONSE_BODY_REGEX, progress);
+                } else if (operator.equalsIgnoreCase("and")) {
+                    responsebodyfound = true;
+                }
+
+                if (operator.equalsIgnoreCase("and")) {
```

Comment:
I must admit I'm finding the logic a bit weird here :)
I dont like the `responsebodyFound = true` when its not _really_ true, just there for the logic.
Suggestion - how about a method like checkPattern(regex, data) which returns a boolean which effectively means pass or fail. Instead of passing in the msg and param you would extract the relevant part. This would replace matchPattern.
Then when you call this from another method like checkMessage(msg) - you could check for failure and if the operator is an 'and' then exit the method straight away with a failure.
Does that make sense?

Reply:
I'm still not happy with this logic ;) I'll try to come up with a better example...

Reply:
How about a method like: `Boolean checkPattern(regex, data)` which would return `null` for no pattern, `true` if it matches otherwise `false`.
Then you can call it at the start of this section like:
```
Boolean requestHeaderFound = checkPattern(this.getData().getRequestHeaderRegex(), msg.getRequestHeader().toString));
Boolean requestBodyFound = checkPattern(this.getData().getRequestBodyRegex(), msg.getRequestBody().toString));
Boolean responseHeaderFound = checkPattern(this.getData().getResponseHeaderRegex(), msg.getResponseHeader().toString));
Boolean responseBodyFound = checkPattern(this.getData().getResponseBodyRegex(), msg.getResponseBody().toString));
```
Then you can do the checks like (pseudocode):
```
if all null
    pass
if "or" and any true
    pass
if "and" and all pass
    pass
fail
```
Does that make sense?
Suggestions from anyone else also appreciated :)